p2panda-net/src/engine/topic_streams.rs

```diff
@@ -0,0 +1,466 @@
+            })?;
+        }
+
+        Ok(())
+    }
+
+    /// Peers exchange topic ids in a process named "topic discovery". This method processes the
+    /// learned topic id's from other peers.
+    pub async fn on_discovered_topic_ids(
+        &mut self,
+        their_topic_ids: Vec<[u8; 32]>,
+        delivered_from: NodeId,
+    ) -> Result<()> {
+        debug!(
+            "learned about topic ids from {}: {:?}",
+            delivered_from, their_topic_ids
+        );
+
+        // Inform the sync manager about any peer-topic combinations which are of interest to us.
+        //
+        // This queues up a sync session which will eventually request the data we are interested
+        // in from that peer.
+        let mut found_common_topic = false;
+        if let Some(sync_actor_tx) = &self.sync_actor_tx {
+            for (topic, _) in self.subscribed.values() {
+                if their_topic_ids.contains(&topic.id()) {
+                    found_common_topic = true;
+                    let peer_topic = ToSyncActor::new(delivered_from, topic.clone());
+                    sync_actor_tx.send(peer_topic).await?
+                }
+            }
+        }
+
+        // Hot path: Optimistically try to join gossip overlays at the same time.
+        if found_common_topic {
+            self.try_join_pending_gossips().await?;
+        }
+
+        Ok(())
+    }
+
+    /// Process new sync session starting with a peer over a topic.
+    #[allow(unused_variables)]
+    pub fn on_sync_start(&self, topic: T, node_id: NodeId) {
+        // Do nothing here for now ..
+    }
+
+    /// Process handshake phase finishing during a sync session.
+    ///
+    /// In the handshake phase peers usually handle authorization and exchange the topic which will
+    /// be synced.
+    pub fn on_sync_handshake_success(&mut self, topic: T, node_id: NodeId) {
+        self.gossip_buffer.lock(node_id, topic.id());
+    }
+
+    /// Process application-data message resulting from the sync session.
+    pub fn on_sync_message(
+        &mut self,
+        topic: T,
+        header: Vec<u8>,
+        payload: Option<Vec<u8>>,
+        delivered_from: NodeId,
+    ) -> Result<()> {
+        let stream_ids = self
+            .topic_to_stream
+            .get(&topic)
+            .expect("consistent topic to stream id mapping");
+
+        for stream_id in stream_ids {
+            let (_, from_network_tx) = self.subscribed.get(stream_id).expect("stream should exist");
+            from_network_tx.send(FromNetwork::SyncMessage {
+                header: header.clone(),
+                payload: payload.clone(),
+                delivered_from: to_public_key(delivered_from),
+            })?;
+        }
+
+        Ok(())
+    }
+
+    /// Process sync session finishing.
+    pub async fn on_sync_done(&mut self, topic: T, node_id: NodeId) -> Result<()> {
+        let topic_id = topic.id();
+        let counter = self.gossip_buffer.unlock(node_id, topic_id);
```

Comment:
Just to check I'm understanding the behaviour here:
- a peer subscribes to two topics `A`+`B` which both have the same topic id `1`
- two independent sync sessions kick-in, one for each topic
- one buffer is created and used for both sync sessions (buffer counter == `2`)
- sync session for topic `A` finishes (buffer counter == `1`)
- sync session for topic `B` finishes (buffer counter == `0`)
- buffer counter is `0` so all messages are drained and forwarded to the two existing subscription channels for topic `A`+`B` (because both have topic id `1`)

Reply:
Sounds right!

We don't always need to necessarily block for incoming sync requests when there's no bi-directional data exchange, but since we don't control the actual sync implementation this is the most straight-forward.